### PR TITLE
MRG: EEG Reference, now also in Epochs and Evoked

### PR DESCRIPTION
**Pull request by wmvanvliet**
continuation from #1713

### CODE REVIEW

**Comment by wmvanvliet** [comment]
@dengemann can you give this a try? The referencing function can now also take an Epochs or Evoked object as parameter.

File: examples/plot_rereference_eeg.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+epochs_params = dict(events=events, event_id=event_id, tmin=tmin, tmax=tmax,
+                     picks=picks, reject=reject)
+
+# No reference. This assumes that the EEG has already been referenced properly.
+# This explicitly prevents MNE from adding a default EEG reference.
+raw_no_ref, _ = mne.io.set_eeg_reference(raw, [])
+evoked_no_ref = mne.Epochs(raw_no_ref, **epochs_params).average()
+del raw_no_ref  # Free memory
+
+fig, (ax1, ax2, ax3) = plt.subplots(nrows=3, ncols=1, sharex=True)
```
Comment (dengemann) [inline examples/plot_rereference_eeg.py]: leave one line empty below for readanility

File: mne/io/reference.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+def set_eeg_reference(data, ref_channels=None, copy=True):
+    """Rereference EEG channels to new reference channel(s).
+
+    If multiple reference channels are specified, they will be averaged. If
+    no reference channels are specified, an average reference will be applied.
+
+    Parameters
+    ----------
+    data : instance of Raw | Epochs | Evoked
```
Comment (dengemann) [inline mne/io/reference.py]: should be `inst` by convention -- `data` is typically used for numpy arrays.

File: mne/io/reference.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    1st cathode, the 2nd anode from the 2nd cathode, etc.
+
+    By default, the virtual channels will be annotated with channel info of
+    the anodes, their locations set to (0, 0, 0) and coil types set to
+    EEG_BIPOLAR.
+
+    Parameters
+    ----------
+    data : instance of Raw | Epochs |Evoked
+        Data containing the unreferenced channels.
```
Comment (dengemann) [inline mne/io/reference.py]: same here.

File: mne/io/tests/test_reference.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+warnings.simplefilter('always')  # enable b/c these tests throw warnings
+
+data_dir = op.join(testing.data_path(download=False), 'MEG', 'sample')
+fif_fname = op.join(data_dir, 'sample_audvis_trunc_raw.fif')
+eve_fname = op.join(data_dir, 'sample_audvis_trunc_raw-eve.fif')
+ave_fname = op.join(data_dir, 'sample_audvis_trunc-ave.fif')
+
+def _test_reference(raw, reref, ref_data, ref_from):
+    '''Helper function to test whether a reference has been correctly
+    applied'''
```
Comment (dengemann) [inline mne/io/tests/test_reference.py]: normally we use " not ' for doc strings

File: mne/io/tests/test_reference.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    raw_other_data = _data[..., picks_other, :]
+
+    # Get the rereferenced EEG data
+    reref_eeg_data = _reref[..., picks_eeg, :]
+    reref_other_data = _reref[..., picks_other, :]
+
+    # Undo rereferencing of EEG channels
+    if isinstance(raw, Epochs):
+        unref_eeg_data = reref_eeg_data + np.tile(ref_data[:, np.newaxis, :],
+                                                  (1, len(picks_eeg), 1))
```
Comment (dengemann) [inline mne/io/tests/test_reference.py]: can you just use numpy broadcasting here? tiling is almost never necessary

File: mne/io/tests/test_reference.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    # Check that both EEG data and other data is the same
+    assert_allclose(raw_eeg_data, unref_eeg_data, 1e-6, atol=1e-15)
+    assert_allclose(raw_other_data, reref_other_data, 1e-6, atol=1e-15)
+
+@testing.requires_testing_data
+def test_apply_reference():
+    '''Test base function for rereferencing'''
+    raw = Raw(fif_fname, preload=True, add_eeg_ref=True)
+
+
```
Comment (dengemann) [inline mne/io/tests/test_reference.py]: remove one line

File: mne/io/tests/test_reference.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    # Test that data is modified in place when copy=False
+    reref, ref_data = set_eeg_reference(raw, ['EEG 001', 'EEG 002'],
+                                        copy=False)
+    assert_true(raw is reref)
+
+
+@testing.requires_testing_data
+def test_set_bipolar_reference():
+    '''Test bipolar referencing'''
```
Comment (dengemann) [inline mne/io/tests/test_reference.py]: """

File: mne/minimum_norm/inverse.py [hunk 8]
```diff
@@ -716,6 +717,15 @@ def _check_ori(pick_ori, pick_normal):
     return pick_ori
 
 
+def _check_reference(inst):
```
Comment (dengemann) [inline mne/minimum_norm/inverse.py]: doc comment """Aux function"""

**Comment by dengemann** [comment]
@wmvanvliet feel free to merge after having addressed my comments. Good work!

**Comment by Eric89GXL** [comment]
Please rebase and check Travis before merging, this will ensure docstrings are automatically checked (recent PR merge).

**Comment by wmvanvliet** [comment]
can I press the green button now?

**Comment by dengemann** [comment]
The honor is yours @wmvanvliet  :)

**Comment by dengemann** [comment]
Thanks for this nice contrib @wmvanvliet! And als for enduring the review sprint and all my nit-picky inputs.

**Comment by agramfort** [comment]
thanks @wmvanvliet !

**Pull request merged**